File: skdltn210/week29/14226.py
from collections import deque

s = int(input())


def bfs(s):
    q = deque([(1, 0, 0)])
    visited = set([(1, 0)])
    while q:
        screen, clipboard, time = q.popleft()
        if screen == s:
            return time
        next = (screen, screen)
        if next not in visited:
            q.append((screen, screen, time + 1))
            visited.add(next)
        if clipboard > 0:
            next = (screen + clipboard, clipboard)
            if next not in visited:
                q.append((screen + clipboard, clipboard, time + 1))
                visited.add(next)
        if screen > 0:
            next = (screen - 1, clipboard)
            if next not in visited:
                q.append((screen - 1, clipboard, time + 1))
                visited.add(next)


print(bfs(s))

File: iam454/week29/14226.js
const fs = require("fs");

const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim();

const S = Number(input);
const MAX_SIZE = 2000;

const visited = Array.from({ length: MAX_SIZE + 1 }, () =>
  new Array(MAX_SIZE + 1).fill(false)
);

const q = [];
q.push([1, 0, 0]);
visited[1][0] = true;

while (q.length) {
  const [screen, clipboard, time] = q.shift();

  if (screen === S) {
    console.log(time);
    break;
  }

  if (visited[screen][screen] === false) {
    visited[screen][screen] = true;
    q.push([screen, screen, time + 1]);
  }

  if (
    clipboard > 0 &&
    screen + clipboard <= MAX_SIZE &&
    visited[screen + clipboard][clipboard] === false
  ) {
    visited[screen + clipboard][clipboard] = true;
    q.push([screen + clipboard, clipboard, time + 1]);
  }

  if (screen > 1 && visited[screen - 1][clipboard] === false) {
    visited[screen - 1][clipboard] = true;
    q.push([screen - 1, clipboard, time + 1]);
  }
}
